File: workspace/src/localization/src/slam_processor.cpp
#include "localization/slam_processor.h"
#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
#include <std_msgs/String.h>

namespace localization {

SLAMProcessor::SLAMProcessor(ros::NodeHandle& nh, ros::NodeHandle& pnh)
    : nh_(nh), pnh_(pnh), pose_initialized_(false) {
    
    // 加载参数
    pnh_.param("filter_alpha", filter_alpha_, 0.8);
    pnh_.param("publish_rate", publish_rate_, 50.0);
    pnh_.param("timeout_threshold", timeout_threshold_, 0.5);
    pnh_.param("map_frame", map_frame_, std::string("map"));
    pnh_.param("base_frame", base_frame_, std::string("base_link"));
    
    // 初始化TF2
    tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>();
    tf_buffer_ = std::make_unique<tf2_ros::Buffer>();
    tf_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf_buffer_);
    
    // 初始化发布器
    pose_pub_ = nh_.advertise<geometry_msgs::PoseStamped>("localization/current_pose", 10);
    odom_pub_ = nh_.advertise<nav_msgs::Odometry>("localization/filtered_odom", 10);
    status_pub_ = nh_.advertise<std_msgs::String>("localization/status", 10);
    
    // 初始化订阅器
    slam_sub_ = nh_.subscribe("/Odometry", 10, &SLAMProcessor::slamCallback, this);
    
    // 初始化定时器
    publish_timer_ = nh_.createTimer(ros::Duration(1.0 / publish_rate_), 
                                   &SLAMProcessor::publishTimerCallback, this);
    
    // 初始化滤波位置
    filtered_pose_.position.x = 0.0;
    filtered_pose_.position.y = 0.0;
    filtered_pose_.position.z = 0.0;
    filtered_pose_.orientation.w = 1.0;
    filtered_pose_.orientation.x = 0.0;
    filtered_pose_.orientation.y = 0.0;
    filtered_pose_.orientation.z = 0.0;
    
    ROS_INFO("📍 SLAM位置处理器已启动");
    ROS_INFO("⚙️  滤波系数: %.2f", filter_alpha_);
    ROS_INFO("📊 发布频率: %.1f Hz", publish_rate_);
    ROS_INFO("🗺️  坐标系: %s -> %s", map_frame_.c_str(), base_frame_.c_str());
}

void SLAMProcessor::run() {
    ROS_INFO("📍 等待SLAM数据...");
    
    // 等待SLAM数据
    while (ros::ok() && !current_odom_) {
        ros::spinOnce();
        ros::Duration(0.1).sleep();
    }
    
    ROS_INFO("✅ SLAM数据已接收，开始位置处理");
    ros::spin();
}

void SLAMProcessor::slamCallback(const nav_msgs::Odometry::ConstPtr& msg) {
    current_odom_ = msg;
    last_update_time_ = msg->header.stamp;
    
    // 滤波处理位置
    filterPose(msg->pose.pose);
    
    // 更新位置历史
    Eigen::Vector3d current_pos = poseToVector3d(filtered_pose_.position);
    updatePositionHistory(msg->header.stamp, current_pos);
}

void SLAMProcessor::filterPose(const geometry_msgs::Pose& new_pose) {
    if (!pose_initialized_) {
        // 初始化
        filtered_pose_ = new_pose;
        pose_initialized_ = true;
        ROS_INFO("📍 位置滤波器已初始化");
    } else {
        // 低通滤波 - 位置
        filtered_pose_.position.x = filter_alpha_ * filtered_pose_.position.x + 
                                   (1.0 - filter_alpha_) * new_pose.position.x;
        filtered_pose_.position.y = filter_alpha_ * filtered_pose_.position.y + 
                                   (1.0 - filter_alpha_) * new_pose.position.y;
        filtered_pose_.position.z = filter_alpha_ * filtered_pose_.position.z + 
                                   (1.0 - filter_alpha_) * new_pose.position.z;
        
        // 姿态直接更新 (可以改为四元数插值)
        filtered_pose_.orientation = new_pose.orientation;
    }
}

void SLAMProcessor::publishTimerCallback(const ros::TimerEvent& event) {
    if (!current_odom_ || !pose_initialized_) {
        return;
    }
    
    ros::Time current_time = ros::Time::now();
    
    // 检查数据时效性
    if ((current_time - last_update_time_).toSec() > timeout_threshold_) {
        ROS_WARN_THROTTLE(5, "⚠️  SLAM数据过期 (%.2f秒)", 
                         (current_time - last_update_time_).toSec());
        return;
    }
    
    // 发布位置和TF
    publishPose(current_time);
    publishTransform(current_time);
    
    // 发布状态
    std_msgs::String status_msg;
    status_msg.data = "HEALTHY";
    status_pub_.publish(status_msg);
    
    // 定期日志
    static int log_counter = 0;
    if (++log_counter % static_cast<int>(publish_rate_ * 5) == 0) {
        ROS_INFO("📍 当前位置: [%.2f, %.2f, %.2f]", 
                filtered_pose_.position.x, 
                filtered_pose_.position.y, 
                filtered_pose_.position.z);
    }
}

void SLAMProcessor::publishPose(const ros::Time& stamp) {
    // 发布当前位置 (给规划模块)
    geometry_msgs::PoseStamped pose_msg;
    pose_msg.header.stamp = stamp;
    pose_msg.header.frame_id = map_frame_;
    pose_msg.pose = filtered_pose_;
    pose_pub_.publish(pose_msg);
    
    // 发布滤波后的里程计 (给可视化和控制)
    nav_msgs::Odometry odom_msg;
    odom_msg.header.stamp = stamp;
    odom_msg.header.frame_id = map_frame_;
    odom_msg.child_frame_id = base_frame_;
    odom_msg.pose.pose = filtered_pose_;
    
    // 估计速度
    Eigen::Vector3d velocity = estimateVelocity();
    odom_msg.twist.twist.linear.x = velocity.x();
    odom_msg.twist.twist.linear.y = velocity.y();
    odom_msg.twist.twist.linear.z = velocity.z();
    
    // 复制原始协方差 (如果有)
    if (current_odom_->pose.covariance[0] > 0) {
        odom_msg.pose.covariance = current_odom_->pose.covariance;
    }
    
    odom_pub_.publish(odom_msg);
}

void SLAMProcessor::publishTransform(const ros::Time& stamp) {
    geometry_msgs::TransformStamped transform;
    transform.header.stamp = stamp;
    transform.header.frame_id = map_frame_;
    transform.child_frame_id = base_frame_;
    
    transform.transform.translation.x = filtered_pose_.position.x;
    transform.transform.translation.y = filtered_pose_.position.y;
    transform.transform.translation.z = filtered_pose_.position.z;
    
    transform.transform.rotation = filtered_pose_.orientation;
    
    tf_broadcaster_->sendTransform(transform);
}

void SLAMProcessor::updatePositionHistory(const ros::Time& stamp, const Eigen::Vector3d& position) {
    position_history_.emplace_back(stamp, position);
    
    // 限制历史大小
    while (position_history_.size() > MAX_HISTORY_SIZE) {
        position_history_.pop_front();
    }
}

Eigen::Vector3d SLAMProcessor::estimateVelocity() const {
    if (position_history_.size() < 2) {
        return Eigen::Vector3d::Zero();
    }
    
    // 使用最近两个位置估计速度
    const auto& latest = position_history_.back();
    const auto& previous = position_history_[position_history_.size() - 2];
    
    double dt = (latest.first - previous.first).toSec();
    if (dt <= 0) {
        return Eigen::Vector3d::Zero();
    }
    
    return (latest.second - previous.second) / dt;
}

geometry_msgs::PoseStamped SLAMProcessor::getCurrentPose() const {
    geometry_msgs::PoseStamped pose;
    pose.header.stamp = ros::Time::now();
    pose.header.frame_id = map_frame_;
    pose.pose = filtered_pose_;
    return pose;
}

bool SLAMProcessor::isLocalizationHealthy() const {
    if (!current_odom_ || !pose_initialized_) {
        return false;
    }
    
    double time_since_update = (ros::Time::now() - last_update_time_).toSec();
    return time_since_update < timeout_threshold_;
}

// 工具函数实现
Eigen::Vector3d SLAMProcessor::poseToVector3d(const geometry_msgs::Point& point) {
    return Eigen::Vector3d(point.x, point.y, point.z);
}

geometry_msgs::Point SLAMProcessor::vector3dToPoint(const Eigen::Vector3d& vec) {
    geometry_msgs::Point point;
    point.x = vec.x();
    point.y = vec.y();
    point.z = vec.z();
    return point;
}

} // namespace localization 

File: workspace/src/localization/scripts/slam_processor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
SLAM位置处理器

功能: 处理LIVO SLAM输出，提供标准化的位置信息给规划和控制模块
"""

import rospy
import numpy as np
from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseStamped, TransformStamped
import tf2_ros
from tf2_geometry_msgs import do_transform_pose

class SLAMProcessor:
    def __init__(self):
        rospy.init_node('slam_processor', anonymous=True)
        
        # 参数配置
        self.filter_alpha = rospy.get_param('~filter_alpha', 0.8)  # 滤波系数
        self.publish_rate = rospy.get_param('~publish_rate', 50.0)  # 发布频率
        
        # 状态变量
        self.current_odom = None
        self.filtered_pose = None
        self.last_update_time = rospy.Time(0)
        
        # TF2
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
        self.tf_broadcaster = tf2_ros.TransformBroadcaster()
        
        # 发布器
        self.pose_pub = rospy.Publisher(
            'current_pose', 
            PoseStamped, 
            queue_size=10
        )
        
        self.odom_pub = rospy.Publisher(
            'filtered_odom',
            Odometry,
            queue_size=10
        )
        
        # 订阅器 - LIVO SLAM输出
        self.slam_sub = rospy.Subscriber(
            '/Odometry',  # LIVO输出
            Odometry,
            self.slam_callback
        )
        
        # 定时发布
        self.timer = rospy.Timer(
            rospy.Duration(1.0/self.publish_rate),
            self.publish_pose
        )
        
        rospy.loginfo("📍 SLAM位置处理器已启动")
        rospy.loginfo(f"⚙️  滤波系数: {self.filter_alpha}")
        rospy.loginfo(f"📊 发布频率: {self.publish_rate} Hz")
    
    def slam_callback(self, msg):
        """接收LIVO SLAM位置数据"""
        self.current_odom = msg
        self.last_update_time = msg.header.stamp
        
        # 简单低通滤波
        if self.filtered_pose is None:
            self.filtered_pose = msg.pose.pose
        else:
            # 位置滤波
            self.filtered_pose.position.x = (
                self.filter_alpha * self.filtered_pose.position.x +
                (1 - self.filter_alpha) * msg.pose.pose.position.x
            )
            self.filtered_pose.position.y = (
                self.filter_alpha * self.filtered_pose.position.y +
                (1 - self.filter_alpha) * msg.pose.pose.position.y
            )
            self.filtered_pose.position.z = (
                self.filter_alpha * self.filtered_pose.position.z +
                (1 - self.filter_alpha) * msg.pose.pose.position.z
            )
            
            # 姿态更新 (可以更复杂的四元数滤波)
            self.filtered_pose.orientation = msg.pose.pose.orientation
    
    def publish_pose(self, event):
        """发布标准化位置信息"""
        if self.current_odom is None or self.filtered_pose is None:
            return
        
        current_time = rospy.Time.now()
        
        # 检查数据时效性
        if (current_time - self.last_update_time).to_sec() > 0.5:
            rospy.logwarn_throttle(5, "⚠️  SLAM数据过期")
            return
        
        # 发布当前位置 (给规划模块)
        pose_msg = PoseStamped()
        pose_msg.header.stamp = current_time
        pose_msg.header.frame_id = "map"
        pose_msg.pose = self.filtered_pose
        self.pose_pub.publish(pose_msg)
        
        # 发布滤波后的里程计 (给可视化)
        odom_msg = Odometry()
        odom_msg.header.stamp = current_time
        odom_msg.header.frame_id = "map"
        odom_msg.child_frame_id = "base_link"
        odom_msg.pose.pose = self.filtered_pose
        if self.current_odom.twist:
            odom_msg.twist = self.current_odom.twist
        self.odom_pub.publish(odom_msg)
        
        # 发布TF变换
        self.publish_transform(current_time)
        
        # 状态日志
        if rospy.get_time() % 5 < 0.1:
            pos = self.filtered_pose.position
            rospy.loginfo(f"📍 当前位置: [{pos.x:.2f}, {pos.y:.2f}, {pos.z:.2f}]")
    
    def publish_transform(self, stamp):
        """发布坐标变换"""
        t = TransformStamped()
        t.header.stamp = stamp
        t.header.frame_id = "map"
        t.child_frame_id = "base_link"
        
        t.transform.translation.x = self.filtered_pose.position.x
        t.transform.translation.y = self.filtered_pose.position.y
        t.transform.translation.z = self.filtered_pose.position.z
        
        t.transform.rotation = self.filtered_pose.orientation
        
        self.tf_broadcaster.sendTransform(t)
    
    def get_current_pose(self):
        """获取当前位置 (供其他模块调用)"""
        return self.filtered_pose if self.filtered_pose else None
    
    def run(self):
        """运行处理器"""
        rospy.loginfo("📍 等待SLAM数据...")
        
        # 等待SLAM数据
        while not rospy.is_shutdown() and self.current_odom is None:
            rospy.sleep(0.1)
        
        rospy.loginfo("✅ SLAM数据已接收，开始位置处理")
        rospy.spin()

if __name__ == '__main__':
    try:
        processor = SLAMProcessor()
        processor.run()
    except rospy.ROSInterruptException:
        rospy.loginfo("🛑 SLAM处理器关闭")
    except Exception as e:
        rospy.logerr(f"❌ SLAM处理器异常: {e}") 